resmem.c

```diff
@@ -50,37 +52,295 @@ static inline int get_info_infile(char *fname, char *res, void *out)
+ */
+static size_t cgmemlimit(char *cg, char *f)
+{
+	char *copy;
+	char *dir;
+	size_t mem = 0, retmem = 0;
+
+	/* read memory limit for cgroup */
+	if ((retmem = cgmemread(cg, f)) == -1) {
+		return 0;
+	}
+
+	copy = strdup(cg);
+	dir = dirname(copy);
+
+	/*read memory limit for parant cg */
+	if (strcmp(dir, "/") != 0) {
+		if((mem = cgmemread(dir, f)) == 0) {
+			free(copy);
+			return 0;
+		}
+		if (mem < retmem) {
+			retmem = mem;
+		}
+	}
+
+	free(copy);
+	return retmem;
+}
+
+/* get memory stat from memory.stat file */
+static inline size_t cgmemstat(char *cg, char *stat)
+{
+	char fn[FNAMELEN];
+	size_t ret;
+
+	snprintf(fn, FNAMELEN, "%s/%s%s/%s", DEFAULTCGFS,
+		MEMCGNAME, cg, "memory.stat");
+
+	if (get_info_infile(fn, stat, &ret) == -1)
+		return 0;
+
+	return (ret / 1024);
+}
+
+/* Get all memory info for a cgroup */
+static int getmeminfoall(char *cg, void *out)
+{
+	size_t memtotal, mmusage, mmtot, cache, active_anon, inactive_anon,
+		active_file, inactive_file, swaptotal, swapfree, swusage,
+		swtot;
+	char fn[FNAMELEN];
+	FILE *fp;
+	int err;
+	char buf[MEMBUF_128];
+
+	snprintf(fn, FNAMELEN, "%s/%s%s/%s", DEFAULTCGFS,
+		MEMCGNAME, cg, "memory.stat");
+
+	fp = fopen(fn, "r");
+
+	if (fp == NULL) {
+		err = errno;
+		eprintf("while opening File %s with errno: %d", fn, errno);
+		errno = err;
+		return -1;
+	}
+
+	while(fgets(buf, sizeof(buf), fp) != NULL) {
+		if (startswith("cache", buf) == 0) {
+			sscanf(buf, "%*s%zu", &cache);
+			cache /= 1024;
+		} else if (startswith("active_anon", buf) == 0) {
+			sscanf(buf, "%*s%zu", &active_anon);
+			active_anon /= 1024;
+		} else if (startswith("inactive_anon", buf) == 0) {
+			sscanf(buf, "%*s%zu", &inactive_anon);
+			inactive_anon /= 1024;
+		} else if (startswith("active_file", buf) == 0) {
+			sscanf(buf, "%*s%zu", &active_file);
+			active_file /= 1024;
+		} else if (startswith("inactive_file", buf) == 0) {
+			sscanf(buf, "%*s%zu", &inactive_file);
+			inactive_file /= 1024;
+		}
+	}
+	fclose(fp);
+
+	fp = fopen(MEMINFO_FILE, "r");
+	if (fp == NULL) {
+		err = errno;
+		eprintf("while opening File %s with errno: %d",
+			MEMINFO_FILE, errno);
+		errno = err;
+		return -1;
+	}
+
+	while(fgets(buf, sizeof(buf), fp) != NULL) {
+		if (startswith("MemTotal", buf) == 0) {
+			sscanf(buf, "%*s%zu", &memtotal);
+		} else if (startswith("SwapTotal", buf) == 0) {
+			sscanf(buf, "%*s%zu", &swaptotal);
+		} else if (startswith("SwapFree", buf) == 0) {
+			sscanf(buf, "%*s%zu", &swapfree);
+		}
+	}
+	fclose(fp);
+
+	mmusage = cgmemread(cg, "memory.usage_in_bytes");
+	mmtot = cgmemlimit(cg, "memory.limit_in_bytes");
+	if (memtotal < mmtot) {
+		mmtot = memtotal;
+	}
+
+	((res_mem_infoall_t *)out)->memfree = mmtot - mmusage;
+	((res_mem_infoall_t *)out)->memavailable = mmtot - mmusage + cache;
+	((res_mem_infoall_t *)out)->memtotal = mmtot;
+	((res_mem_infoall_t *)out)->active = active_anon + active_file;
+	((res_mem_infoall_t *)out)->inactive = inactive_anon + inactive_file;
```

Comment:
Would it make more sense to have a point on the stack instead of doing multiple casts? Like:
```c
res_mem_infoall_t *memall = out;
...
memall->memfree = mmtot - mmusage;
...
```

Reply:
Makes sense, will change.